### PR TITLE
ch9628: streamset earliest/latest fix

### PR DESCRIPTION
**Pull request by mchestnut91**

### CODE REVIEW

**Comment by mchestnut91** [comment]
The DS team brought it to my attention that `StreamSet.earliest()` and `StreamSet.latest()` queries were taking ~30 minutes to get all of the points for a streamset containing 183 streams.

I took a look and the issue was that `StreamSet.versions()` was issuing a call to `StreamSet._latest_versions()` each time, and never called `StreamSet.pin_versions()` so the versions could actually get pinned for later calls. 

I wasn't exactly sure what to return in `StreamSet.pin_versions`, I decided to go with `self._pinned_versions` rather than `self`, but I am open to other suggestions for how to handle this.

File: btrdb/stream.py [hunk 1]
```diff
@@ -845,7 +845,7 @@ def pin_versions(self, versions=None):
 
 
         self._pinned_versions = self._latest_versions() if not versions else versions
-        return self
+        return self._pinned_versions
```
Comment (looselycoupled) [inline btrdb/stream.py:848]: This breaks the concept of method chaining.  Streamset methods like this should always return `self`.

File: btrdb/stream.py [hunk 2]
```diff
@@ -864,7 +864,7 @@ def versions(self):
             A dict containing the stream UUID and version ints as key/values
 
         """
-        return self._pinned_versions if self._pinned_versions else self._latest_versions()
+        return self._pinned_versions if self._pinned_versions else self.pin_versions()
```
Comment (looselycoupled) [inline btrdb/stream.py:867]: Related to above.

File: btrdb/stream.py (hunk 1 above)
Comment (looselycoupled) [reply btrdb/stream.py:848]: I need to adjust my brain to think about the problem you are running into...  will return with comments but I dont think this is the correct fix in this case.

**Comment by mchestnut91** [comment]
@looselycoupled I took another look at this after our conversation last week, and I believe that the approach that you mentioned during our call is the best fix - in the `StreamSet.earliest()` and `StreamSet.latest()` functions, we should be calling `self.versions()` once. This will return a dict of {UUID: version}, which we can use to look up the correct version for each stream by UUID. 

Let me know if you think I'm missing something. I tested it out and the earliest/latest calls are down to ~30 seconds for a StreamSet with 183 streams

**Review by looselycoupled — APPROVED** [review]
LGTM

**Pull request merged**